<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>scrattch.hicat • scrattch.hicat</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="scrattch.hicat">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scrattch.hicat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.23</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/scrattch.hicat.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>scrattch.hicat</h1>
                        <h4 class="author">Zizhen Yao</h4>
                        <h4 class="author">Lucas Graybuck</h4>
            
            <h4 class="date">2019-06-30</h4>
      
      
      <div class="hidden name"><code>scrattch.hicat.Rmd</code></div>

    </div>

    
    
<div id="overview-of-hierarchical-iterative-clustering-for-analysis-of-transcriptomics" class="section level1">
<h1 class="hasAnchor">
<a href="#overview-of-hierarchical-iterative-clustering-for-analysis-of-transcriptomics" class="anchor"></a>Overview of Hierarchical Iterative Clustering for Analysis of Transcriptomics</h1>
<p>The <code>scrattch.hicat</code> package offers functions to perform iterative clustering for single cell RNAseq datasets.</p>
<p>The pipeline consists of the following key steps:<br>
* [High variance gene selection]<br>
* [Dimensionality reduction]<br>
* <a href="#filter">Dimension filtering based on QC correlation</a><br>
* [Jaccard-Louvain or hierarchichal (Ward) clustering]<br>
* <a href="#merge">Cluster merging based on presence of differentially expressed genes.</a></p>
<p>This process is iteratively repeated within each resulting cluster until no more clusters meet differential gene expression and cluster size termination criteria.<br>
We can inspect and visualize the results of consensus clustering:<br>
* <a href="#markerplot">Compute marker genes and plot marker heatmaps</a><br>
* <a href="#dendplot">Hierarchical dendrograms</a><br>
* <a href="#tsneplot">t-SNE plots</a></p>
<p>For this vignette, we use a subset of the dataset published in <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4985242/">Tasic, et al. (2016) Nature Neuroscience</a>, which is available in the <a href="https://github.com/AllenInstitute/tasic2016data">tasic2016data package</a>:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="cf">if</span>(<span class="op">!</span><span class="st">"tasic2016data"</span> <span class="op">%in%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/installed.packages">installed.packages</a></span>())) {</a>
<a class="sourceLine" id="cb1-2" title="2">  devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"AllenInstitute/tasic2016data"</span>)</a>
<a class="sourceLine" id="cb1-3" title="3">}</a>
<a class="sourceLine" id="cb1-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tasic2016data)</a></code></pre></div>
<p>This vignette depends on a few other packages, in addition to <code>scrattch.hicat</code>:</p>
<div id="dataset-formatting-and-setup" class="section level2">
<h2 class="hasAnchor">
<a href="#dataset-formatting-and-setup" class="anchor"></a>Dataset Formatting and Setup</h2>
<p><a id="setup"></a></p>
<p>First prepare the datasets</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="co"># Load sample annotations (anno)</span></a>
<a class="sourceLine" id="cb2-2" title="2">anno &lt;-<span class="st"> </span>tasic_<span class="dv">2016</span>_anno</a>
<a class="sourceLine" id="cb2-3" title="3"></a>
<a class="sourceLine" id="cb2-4" title="4"><span class="co"># Make a data.frame of unique cluster id, type, color, and broad type</span></a>
<a class="sourceLine" id="cb2-5" title="5">ref.cl.df &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/as.data.frame">as.data.frame</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(anno[,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"primary_type_id"</span>, <span class="st">"primary_type_label"</span>, <span class="st">"primary_type_color"</span>, <span class="st">"broad_type"</span>)]))</a>
<a class="sourceLine" id="cb2-6" title="6"></a>
<a class="sourceLine" id="cb2-7" title="7"><span class="co">#standardize cluster annoation with cluster_id, cluster_label and cluster_color. These are the required fields to visualize clusters properly.</span></a>
<a class="sourceLine" id="cb2-8" title="8"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(ref.cl.df)[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>] &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"cluster_id"</span>, <span class="st">"cluster_label"</span>, <span class="st">"cluster_color"</span>)</a>
<a class="sourceLine" id="cb2-9" title="9"></a>
<a class="sourceLine" id="cb2-10" title="10"><span class="co"># Sort by cluster_id</span></a>
<a class="sourceLine" id="cb2-11" title="11">ref.cl.df &lt;-<span class="st"> </span>ref.cl.df[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/order">order</a></span>(ref.cl.df<span class="op">$</span>cluster_id),]</a>
<a class="sourceLine" id="cb2-12" title="12"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/row.names">row.names</a></span>(ref.cl.df) &lt;-<span class="st"> </span>ref.cl.df<span class="op">$</span>cluster_id</a>
<a class="sourceLine" id="cb2-13" title="13"></a>
<a class="sourceLine" id="cb2-14" title="14">ref.cl &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/setNames">setNames</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(anno<span class="op">$</span>primary_type_id), anno<span class="op">$</span>sample_name)</a></code></pre></div>
<p>Convert counts to CPM and take log2 transformation</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1">norm.dat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/logCPM.html">logCPM</a></span>(tasic_<span class="dv">2016</span>_counts)</a></code></pre></div>
<p>If you have a very large matrix, we recommend you convert it to a sparse matrix using the Matrix package to save memory:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1">sparse_counts &lt;-<span class="st"> </span><span class="kw">as</span>(tasic_<span class="dv">2016</span>_counts, <span class="st">"dgCMatrix"</span>)</a>
<a class="sourceLine" id="cb4-2" title="2"></a>
<a class="sourceLine" id="cb4-3" title="3">norm.dat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/logCPM.html">logCPM</a></span>(sparse_counts)</a></code></pre></div>
<p>For this demo, we’ll select a small subset of CGE-derived interneurons for clustering. This gives us a set of 284 single-cell transcriptomic profiles to work with.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1">classified_cells &lt;-<span class="st"> </span>anno<span class="op">$</span>primary_type_label <span class="op">!=</span><span class="st"> "unclassified"</span></a>
<a class="sourceLine" id="cb5-2" title="2">selected_types &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grepl</a></span>(<span class="st">"Igtp|Ndnf|Vip|Sncg|Smad3"</span>, anno<span class="op">$</span>primary_type_label)</a>
<a class="sourceLine" id="cb5-3" title="3"></a>
<a class="sourceLine" id="cb5-4" title="4">select_logic &lt;-<span class="st"> </span>classified_cells <span class="op">&amp;</span><span class="st"> </span>selected_types</a>
<a class="sourceLine" id="cb5-5" title="5">select.cells &lt;-<span class="st"> </span>anno<span class="op">$</span>sample_name[select_logic]</a></code></pre></div>
</div>
<div id="parameter-specification" class="section level2">
<h2 class="hasAnchor">
<a href="#parameter-specification" class="anchor"></a>Parameter specification</h2>
<p><a id="params"></a></p>
<p>The final number of clusters produced by this iterative clustering algorithm is largely determined by the required cell type resolution specified by the user. The cell type resolution is defined by differential expression (DE) criteria between every pair of clusters. The users can specify these criteria ahead of time, for reuse in hicat functions by using the <code><a href="../reference/de_param.html">de_param()</a></code> function.</p>
<p>We compute statistical singificance of DE genes using <code>limma</code>, with two key parameters specified below:<br><strong>padj.th</strong>: adjusted p value threshold for DE genes.<br><strong>lfc.th</strong>: log2 fold change threshold for DE genes.</p>
<p>We also require DE genes to have a relatively binary (on/off) expression pattern, specified by the following parameters:<br><strong>low.th</strong>: The minimum value used to determine whether a gene is detected in a given cell or not. This threshold is applied to log2-transformed, normalized data. The default value is 1. Users can specifiy different thresholds for different genes if necessary.<br>
For every pair of clusters (one as foreground, and the other as background), we define q1, and q2 as the proportion of cells with expression &gt; <strong>low.th</strong> in the foregound and background cluster respectively.<br><strong>q1.th</strong>: For up regulated genes, q1 should be greater than q1.th in the <em>foreground</em> set.<br><strong>q2.th</strong>: For up regulated genes, q2 should be smaller than q2.th in the <em>background</em> set.<br><strong>q.diff.th</strong>: The difference, defined as abs(q1 - q2)/max(q1, q2) should be greater than q.diff.th.<br>
By default, <strong>q1.th</strong> = 0.5, <strong>q2.th</strong> = NULL, and <strong>q.diff.th </strong>= 0.7.</p>
<p>The user can also ignore these parameters by setting them all to NULL.<br>
For high-depth datasets, like those generated using SMARTerV4 or Smart-Seq2, we recommend starting with <strong>q1.th</strong> = 0.5 .<br>
For low-depth datasets, like those generated using Dropseq or 10X Genomics, we recommend starting with <strong>q1.th</strong> = 0.3 due to the generally lower gene detection per cell in these datasets.<br>
When focusing on discrete cell types, set <strong>q.diff.th</strong> closer to 1.<br>
If splitting cell types based on graded gene expression variation is of interest, adjust <strong>q.diff.th</strong> closer to 0.</p>
<p>To determine whether two clusters are seperable based on DE genes, we define <strong>de.score</strong> as the sum of -log10(adjusted Pvalue) for all DE genes. Each gene contributes at most 20 towards the sum. All clusters should have pairwise de.score greater than <strong>de.score.th</strong>.<br>
For small datasets (#cells &lt; 1000), we recommend <strong>de.score.th</strong> = 40.<br>
For large datasets (#cells &gt; 10000), we recommend <strong>de.score.th</strong> = 150.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1">de.param &lt;-<span class="st"> </span><span class="kw"><a href="../reference/de_param.html">de_param</a></span>(<span class="dt">padj.th     =</span> <span class="fl">0.05</span>, </a>
<a class="sourceLine" id="cb6-2" title="2">                     <span class="dt">lfc.th      =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb6-3" title="3">                     <span class="dt">low.th      =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb6-4" title="4">                     <span class="dt">q1.th       =</span> <span class="fl">0.5</span>, </a>
<a class="sourceLine" id="cb6-5" title="5">                     <span class="dt">q.diff.th   =</span> <span class="fl">0.7</span>, </a>
<a class="sourceLine" id="cb6-6" title="6">                     <span class="dt">de.score.th =</span> <span class="dv">40</span>)</a></code></pre></div>
<p>##Perform clustering<br><a id="clust"></a></p>
<p><code>scrattch.hicat</code> can perform clustering using WGCNA or PCA for dimensionality reduction. WGCNA mode is good for detecting rare clusters and provides cleaner cluster boundaries, while PCA is more scalable to large datasets, captures combinatorial marker expression patterns more effectively, and is more sensitive to low-depth datasets.<br>
We recommend using WGCNA for smaller, high-depth datasets (&lt; 4,000 samples; &gt; 5,000 genes detecter per sample), and PCA for large or low-coverage datasets (&gt; 4,000 samples or &lt; 5,000 genes detected per sample). Another consideration is that WGCNA is considerably slower than PCA. Note that while the whole clustering pipeline scales quite well with the number of cells, the running time heavily depends on the cell type complexity as clustering is iterative.</p>
<p>First, let us just run one round of clustering using PCA mode using high stringency to check the broad cell types</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">broad.param &lt;-<span class="st"> </span><span class="kw"><a href="../reference/de_param.html">de_param</a></span>(<span class="dt">padj.th     =</span> <span class="fl">0.05</span>, </a>
<a class="sourceLine" id="cb7-2" title="2">                        <span class="dt">lfc.th      =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb7-3" title="3">                        <span class="dt">low.th      =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb7-4" title="4">                        <span class="dt">q1.th       =</span> <span class="fl">0.5</span>, </a>
<a class="sourceLine" id="cb7-5" title="5">                        <span class="dt">q.diff.th   =</span> <span class="fl">0.7</span>, </a>
<a class="sourceLine" id="cb7-6" title="6">                        <span class="dt">de.score.th =</span> <span class="dv">500</span>)</a>
<a class="sourceLine" id="cb7-7" title="7"></a>
<a class="sourceLine" id="cb7-8" title="8">onestep.result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/onestep_clust.html">onestep_clust</a></span>(norm.dat, </a>
<a class="sourceLine" id="cb7-9" title="9">                                <span class="dt">select.cells =</span> select.cells, </a>
<a class="sourceLine" id="cb7-10" title="10">                                <span class="dt">dim.method =</span> <span class="st">"pca"</span>, </a>
<a class="sourceLine" id="cb7-11" title="11">                                <span class="dt">de.param =</span> broad.param)</a></code></pre></div>
<p><a id="markerplot"></a></p>
<p>Take a quick look at the clustering heatmap</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1">display.result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/display_cl.html">display_cl</a></span>(<span class="dt">cl =</span> onestep.result<span class="op">$</span>cl, </a>
<a class="sourceLine" id="cb8-2" title="2">                             <span class="dt">norm.dat =</span> norm.dat, </a>
<a class="sourceLine" id="cb8-3" title="3">                             <span class="dt">plot =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb8-4" title="4">                             <span class="dt">de.param =</span> de.param)</a></code></pre></div>
<p><img src="scrattch.hicat_files/figure-html/unnamed-chunk-1-1.png" width="672"></p>
<p>Apply iterative clustering pipeline for finer splits based on existing clusters using a more relaxed threshold.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1">PCA.clust.result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/iter_clust.html">iter_clust</a></span>(norm.dat, </a>
<a class="sourceLine" id="cb9-2" title="2">                               <span class="dt">select.cells =</span> select.cells, </a>
<a class="sourceLine" id="cb9-3" title="3">                               <span class="dt">dim.method =</span> <span class="st">"pca"</span>, </a>
<a class="sourceLine" id="cb9-4" title="4">                               <span class="dt">de.param =</span> de.param, </a>
<a class="sourceLine" id="cb9-5" title="5">                               <span class="dt">result =</span> onestep.result)</a></code></pre></div>
<p>You can also apply iterative clustering pipeline from the beginning by specifying substantially stringent clustering requirements and leaving out the <code>result</code> parameter:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1">PCA.clust.result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/iter_clust.html">iter_clust</a></span>(norm.dat, </a>
<a class="sourceLine" id="cb10-2" title="2">                               <span class="dt">select.cells =</span> select.cells, </a>
<a class="sourceLine" id="cb10-3" title="3">                               <span class="dt">dim.method =</span> <span class="st">"pca"</span>, </a>
<a class="sourceLine" id="cb10-4" title="4">                               <span class="dt">de.param =</span> de.param)</a></code></pre></div>
<p>We should get more clusters than in the previous run:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1">display.result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/display_cl.html">display_cl</a></span>(<span class="dt">cl =</span> PCA.clust.result<span class="op">$</span>cl, </a>
<a class="sourceLine" id="cb11-2" title="2">                             <span class="dt">norm.dat =</span> norm.dat, </a>
<a class="sourceLine" id="cb11-3" title="3">                             <span class="dt">plot =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb11-4" title="4">                             <span class="dt">de.param =</span> de.param)</a></code></pre></div>
<p><img src="scrattch.hicat_files/figure-html/unnamed-chunk-4-1.png" width="672"></p>
</div>
<div id="eliminate-technical-artfacts-at-dimension-reduction-state" class="section level2">
<h2 class="hasAnchor">
<a href="#eliminate-technical-artfacts-at-dimension-reduction-state" class="anchor"></a>Eliminate technical artfacts at dimension reduction state</h2>
<p><a id="filter"></a></p>
<p>Technical variation that should be masked during the clustering process can be specified using a matrix that we term <strong>rm.eigen</strong>. If batch effects are present, you can use the first principle component of batch-specific genes as a column in <strong>rm.eigen</strong>.<br>
QC-related factors, such as sequencing depth, gene detection limits, and fraction of reads mapped to transcriptome also tend to correlate with systemetic technical variation in gene expression.<br>
You can directly use proper transformation of these QC-related variables, or use the first principle component of the genes that correlate with these QC factors. The latter approach tends to work better with real datasets. When <strong>rm.eigen</strong> is specified, any reduced-dimension vectors during clustering that have correlation greater than <strong>rm.th</strong> with any columns of <strong>rm.eigen</strong> are ignored during clustering.</p>
<p>We recommend that users to first explore their data by running the clustering pipeline without setting <strong>rm.eigen</strong>. If any batch specific or QC-driven clusters appear to cause problems, the user can create <strong>rm.eigen</strong> as demonstrated below, and rerun the pipeline.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">gene.counts &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">colSums</a></span>(norm.dat <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)</a>
<a class="sourceLine" id="cb12-2" title="2">rm.eigen &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log2</a></span>(gene.counts), <span class="dt">ncol =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb12-3" title="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/row.names">row.names</a></span>(rm.eigen) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(gene.counts)</a>
<a class="sourceLine" id="cb12-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(rm.eigen) &lt;-<span class="st"> "log2GeneCounts"</span></a></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1">PCA.clust.result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/iter_clust.html">iter_clust</a></span>(norm.dat, </a>
<a class="sourceLine" id="cb13-2" title="2">                               <span class="dt">select.cells =</span> select.cells, </a>
<a class="sourceLine" id="cb13-3" title="3">                               <span class="dt">dim.method =</span> <span class="st">"pca"</span>, </a>
<a class="sourceLine" id="cb13-4" title="4">                               <span class="dt">de.param =</span> de.param,</a>
<a class="sourceLine" id="cb13-5" title="5">                               <span class="dt">rm.eigen =</span> rm.eigen)</a></code></pre></div>
</div>
<div id="cluster-merging-based-on-presence-of-differentially-expressed-genes-" class="section level2">
<h2 class="hasAnchor">
<a href="#cluster-merging-based-on-presence-of-differentially-expressed-genes-" class="anchor"></a>Cluster merging based on presence of differentially expressed genes.</h2>
<p><a id="merge"></a></p>
<p>Throughout the iterative clustering process performed by <code><a href="../reference/iter_clust.html">iter_clust()</a></code>, the function checked whether clusters at any iteration can be seperated by DEG. However, clusters from different iterations can end up very similar. Thus, it is necessary to check whether all the clusters are seperable by DEGs in the end. Clusters are merged in an order defined by the pairs of clusters that are nearest neighbors, which are computed in a reduced dimension space defined by <strong>rd.dat</strong>.<br>
Here, we use the set of markers produced by <code><a href="../reference/iter_clust.html">iter_clust()</a></code> to define the reduced dimensions:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1">marker.dat &lt;-<span class="st"> </span>norm.dat[PCA.clust.result<span class="op">$</span>markers, select.cells]</a>
<a class="sourceLine" id="cb14-2" title="2"></a>
<a class="sourceLine" id="cb14-3" title="3">PCA.merge.result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/merge_cl.html">merge_cl</a></span>(norm.dat, </a>
<a class="sourceLine" id="cb14-4" title="4">                             <span class="dt">cl =</span> PCA.clust.result<span class="op">$</span>cl, </a>
<a class="sourceLine" id="cb14-5" title="5">                             <span class="dt">rd.dat =</span> marker.dat,</a>
<a class="sourceLine" id="cb14-6" title="6">                             <span class="dt">de.param =</span> de.param)</a></code></pre></div>
</div>
<div id="compare-and-annotate-clusters-against-reference-cluster-annotation-" class="section level2">
<h2 class="hasAnchor">
<a href="#compare-and-annotate-clusters-against-reference-cluster-annotation-" class="anchor"></a>compare and annotate clusters against reference cluster annotation.</h2>
<p><a id="annotate"></a></p>
<p>In this demo, we have previously derived cluster labels from Tasic, et al. (2016). To see how these cluster calls match up to those generated by <code>scrattch.hicat</code>, we can compare and annotate the clusters based on the reference cluster annotation. <img src="scrattch.hicat_files/figure-html/Compare%20to%20benchmark-1.png" width="624"></p>
<p>Now Let’s compute DE genes between every pair of clusters. To generate the heatmap, use the top 20 DE genes b default. If you have many clusters, the heatmap can be too large to display. The users may choose different “n.markers”</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1">display.result =<span class="st"> </span><span class="kw"><a href="../reference/display_cl.html">display_cl</a></span>(cl, </a>
<a class="sourceLine" id="cb15-2" title="2">                            norm.dat, </a>
<a class="sourceLine" id="cb15-3" title="3">                            <span class="dt">plot =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb15-4" title="4">                            <span class="dt">de.param =</span> de.param, </a>
<a class="sourceLine" id="cb15-5" title="5">                            <span class="dt">min.sep =</span> <span class="dv">4</span>, </a>
<a class="sourceLine" id="cb15-6" title="6">                            <span class="dt">n.markers =</span> <span class="dv">20</span>)</a></code></pre></div>
<p><img src="scrattch.hicat_files/figure-html/unnamed-chunk-7-1.png" width="672"></p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1">de.genes &lt;-<span class="st"> </span>display.result<span class="op">$</span>de.genes</a></code></pre></div>
<p>At this point, check the clusters manually to determine if some clusters are outliers. Define cl.clean as the clusters after removing outliers.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1">cl.clean &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/droplevels">droplevels</a></span>(cl)</a></code></pre></div>
<p><a id="dendplot"></a> We find it helpful to use hierarchical structure to categorize cell types at different resolution. To build dendrogram, we use cluster-cluster correlation matrix based on cluster medians of the top 50 genes between every pair of clusters with <code><a href="../reference/select_markers.html">select_markers()</a></code> and <code>build_dend()</code>. The confidence of each branch point can be estimated by bootstrap approach implemented by pvclust package, which is encapsulated by “build_dend” function.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1">select.markers &lt;-<span class="st"> </span><span class="kw"><a href="../reference/select_markers.html">select_markers</a></span>(norm.dat, </a>
<a class="sourceLine" id="cb18-2" title="2">                                 cl.clean, </a>
<a class="sourceLine" id="cb18-3" title="3">                                 <span class="dt">de.genes =</span> de.genes,</a>
<a class="sourceLine" id="cb18-4" title="4">                                 <span class="dt">n.markers =</span> <span class="dv">50</span>)<span class="op">$</span>markers</a>
<a class="sourceLine" id="cb18-5" title="5"></a>
<a class="sourceLine" id="cb18-6" title="6">marker.dat &lt;-<span class="st"> </span>norm.dat[select.markers, ]</a>
<a class="sourceLine" id="cb18-7" title="7"></a>
<a class="sourceLine" id="cb18-8" title="8">cl.med &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_cl_medians.html">get_cl_medians</a></span>(marker.dat, cl)</a>
<a class="sourceLine" id="cb18-9" title="9"></a>
<a class="sourceLine" id="cb18-10" title="10"><span class="co">## The prefered order for the leaf nodes.</span></a>
<a class="sourceLine" id="cb18-11" title="11">l.rank &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/setNames">setNames</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(cl.df), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/row.names">row.names</a></span>(cl.df))</a>
<a class="sourceLine" id="cb18-12" title="12"><span class="co">## Color of the leaf nodes.</span></a>
<a class="sourceLine" id="cb18-13" title="13">l.color &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/setNames">setNames</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(cl.df<span class="op">$</span>cluster_color), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/row.names">row.names</a></span>(cl.df))</a>
<a class="sourceLine" id="cb18-14" title="14"></a>
<a class="sourceLine" id="cb18-15" title="15">dend.result &lt;-<span class="st"> </span><span class="kw">build_dend</span>(cl.med[,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/levels">levels</a></span>(cl.clean)],</a>
<a class="sourceLine" id="cb18-16" title="16">                          l.rank, </a>
<a class="sourceLine" id="cb18-17" title="17">                          l.color,</a>
<a class="sourceLine" id="cb18-18" title="18">                          <span class="dt">nboot =</span> <span class="dv">100</span>)</a>
<a class="sourceLine" id="cb18-19" title="19"></a>
<a class="sourceLine" id="cb18-20" title="20">dend &lt;-<span class="st"> </span>dend.result<span class="op">$</span>dend</a>
<a class="sourceLine" id="cb18-21" title="21"></a>
<a class="sourceLine" id="cb18-22" title="22"><span class="co">###attach cluster labels to the leafs of the tree </span></a>
<a class="sourceLine" id="cb18-23" title="23">dend.labeled &lt;-<span class="st"> </span>dend</a>
<a class="sourceLine" id="cb18-24" title="24"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/labels">labels</a></span>(dend.labeled) &lt;-<span class="st"> </span>cl.df[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/labels">labels</a></span>(dend), <span class="st">"cluster_label"</span>]</a>
<a class="sourceLine" id="cb18-25" title="25"></a>
<a class="sourceLine" id="cb18-26" title="26"><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(dend.labeled)</a></code></pre></div>
<p><img src="scrattch.hicat_files/figure-html/Build%20dendrogram-1.png" width="672"></p>
<p>Reorder the clusters based on the dendrogram:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1">cl.clean &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/setNames">setNames</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(cl.clean), <span class="dt">levels =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/labels">labels</a></span>(dend)), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(cl.clean))</a>
<a class="sourceLine" id="cb19-2" title="2">cl.df.clean &lt;-<span class="st"> </span>cl.df[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/levels">levels</a></span>(cl.clean),]</a></code></pre></div>
<p>Plot cluster-cluster correlation matrix:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" title="1">cl.cor &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/cor">cor</a></span>(cl.med)</a>
<a class="sourceLine" id="cb20-2" title="2"></a>
<a class="sourceLine" id="cb20-3" title="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(cl.cor) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(cl.cor) &lt;-<span class="st"> </span>cl.df[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/row.names">row.names</a></span>(cl.cor), <span class="st">"cluster_label"</span>]</a>
<a class="sourceLine" id="cb20-4" title="4"><span class="kw"><a href="../reference/heatmap.3.html">heatmap.3</a></span>(cl.cor,</a>
<a class="sourceLine" id="cb20-5" title="5">          <span class="dt">Rowv =</span> dend, </a>
<a class="sourceLine" id="cb20-6" title="6">          <span class="dt">Colv =</span> dend,</a>
<a class="sourceLine" id="cb20-7" title="7">          <span class="dt">trace =</span> <span class="st">"none"</span>, </a>
<a class="sourceLine" id="cb20-8" title="8">          <span class="dt">col =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/grDevices/topics/palettes">heat.colors</a></span>(<span class="dv">100</span>),</a>
<a class="sourceLine" id="cb20-9" title="9">          <span class="dt">cexRow =</span> <span class="fl">0.8</span>, </a>
<a class="sourceLine" id="cb20-10" title="10">          <span class="dt">cexCol =</span> <span class="fl">0.8</span>,</a>
<a class="sourceLine" id="cb20-11" title="11">          <span class="dt">breaks =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="op">-</span><span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="fl">0.2</span>, <span class="dv">1</span>, <span class="dt">length.out =</span> <span class="dv">99</span>)))</a></code></pre></div>
<p><img src="scrattch.hicat_files/figure-html/Plot%20the%20correlation%20heatmap-1.png" width="480"></p>
</div>
<div id="t-sne-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#t-sne-plots" class="anchor"></a>t-SNE plots</h2>
<p><a id="tsneplot"></a></p>
<p>Generate t-SNE coordinates and a plot using <code>plot_tsne_cl()</code>:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" title="1">tsne.result &lt;-<span class="st"> </span><span class="kw">plot_tsne_cl</span>(norm.dat, </a>
<a class="sourceLine" id="cb21-2" title="2">                            select.markers, </a>
<a class="sourceLine" id="cb21-3" title="3">                            cl, </a>
<a class="sourceLine" id="cb21-4" title="4">                            cl.df, </a>
<a class="sourceLine" id="cb21-5" title="5">                            <span class="dt">fn.size =</span> <span class="dv">5</span>, </a>
<a class="sourceLine" id="cb21-6" title="6">                            <span class="dt">cex =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb21-7" title="7"></a>
<a class="sourceLine" id="cb21-8" title="8">tsne.df &lt;-<span class="st"> </span>tsne.result<span class="op">$</span>tsne.df</a>
<a class="sourceLine" id="cb21-9" title="9"></a>
<a class="sourceLine" id="cb21-10" title="10">tsne.result<span class="op">$</span>g</a></code></pre></div>
<p><img src="scrattch.hicat_files/figure-html/tsne%20plot-1.png" width="576"></p>
<p>Plot expression of a set of genes using the t-SNE coordinates with <code>plot_tsne_gene()</code>:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" title="1">tsne.df &lt;-<span class="st"> </span>tsne.result<span class="op">$</span>tsne.df</a>
<a class="sourceLine" id="cb22-2" title="2"></a>
<a class="sourceLine" id="cb22-3" title="3"><span class="co">## 6330527O06Rik is a synonym for Lamp5, and A930038C07Rik is a synonym for Ndnf</span></a>
<a class="sourceLine" id="cb22-4" title="4">markers &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Vip"</span>,<span class="st">"6330527O06Rik"</span>,<span class="st">"Sncg"</span>,<span class="st">"Cxcl14"</span>,<span class="st">"Gpc3"</span>,<span class="st">"A930038C07Rik"</span>)</a>
<a class="sourceLine" id="cb22-5" title="5">gene.plots &lt;-<span class="st"> </span><span class="kw">plot_tsne_gene</span>(tsne.df, </a>
<a class="sourceLine" id="cb22-6" title="6">                             norm.dat, </a>
<a class="sourceLine" id="cb22-7" title="7">                             markers)</a>
<a class="sourceLine" id="cb22-8" title="8"></a>
<a class="sourceLine" id="cb22-9" title="9"><span class="kw">multiplot</span>(<span class="dt">plotlist =</span> gene.plots, </a>
<a class="sourceLine" id="cb22-10" title="10">          <span class="dt">cols =</span> <span class="dv">3</span>)</a></code></pre></div>
<p><img src="scrattch.hicat_files/figure-html/tSNE%20expression,%20-1.png" width="672"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#overview-of-hierarchical-iterative-clustering-for-analysis-of-transcriptomics">Overview of Hierarchical Iterative Clustering for Analysis of Transcriptomics</a><ul class="nav nav-pills nav-stacked">
<li><a href="#dataset-formatting-and-setup">Dataset Formatting and Setup</a></li>
      <li><a href="#parameter-specification">Parameter specification</a></li>
      <li><a href="#eliminate-technical-artfacts-at-dimension-reduction-state">Eliminate technical artfacts at dimension reduction state</a></li>
      <li><a href="#cluster-merging-based-on-presence-of-differentially-expressed-genes-">Cluster merging based on presence of differentially expressed genes.</a></li>
      <li><a href="#compare-and-annotate-clusters-against-reference-cluster-annotation-">compare and annotate clusters against reference cluster annotation.</a></li>
      <li><a href="#t-sne-plots">t-SNE plots</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Zizhen Yao, Lucas Graybuck, Trygve Bakken, Jeremy Miller, Cindy van Velthoven.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
