<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Set differential expression (DE) parameters for genes and clusters. — de_param • scrattch.hicat</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Set differential expression (DE) parameters for genes and clusters. — de_param" />

<meta property="og:description" content="This function provides a convenient way to manage settings for differential expression tests in scrattch.hicat." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scrattch.hicat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.23</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/scrattch.hicat.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Set differential expression (DE) parameters for genes and clusters.</h1>
    
    <div class="hidden name"><code>de_param.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function provides a convenient way to manage settings for differential expression tests in scrattch.hicat.</p>
    
    </div>

    <pre class="usage"><span class='fu'>de_param</span>(<span class='kw'>low.th</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>padj.th</span> <span class='kw'>=</span> <span class='fl'>0.01</span>, <span class='kw'>lfc.th</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>q1.th</span> <span class='kw'>=</span> <span class='fl'>0.5</span>,
  <span class='kw'>q2.th</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>q.diff.th</span> <span class='kw'>=</span> <span class='fl'>0.7</span>, <span class='kw'>de.score.th</span> <span class='kw'>=</span> <span class='fl'>150</span>, <span class='kw'>min.cells</span> <span class='kw'>=</span> <span class='fl'>4</span>,
  <span class='kw'>min.genes</span> <span class='kw'>=</span> <span class='fl'>5</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>low.th</th>
      <td><p>Lower boundary for normalized gene expression. Default = 1. See details.</p></td>
    </tr>
    <tr>
      <th>padj.th</th>
      <td><p>Upper boundary threshold for adjusted p-values for differential expression tests. Default = 0.01. See details.</p></td>
    </tr>
    <tr>
      <th>lfc.th</th>
      <td><p>Lower boundary threshold for log2(fold change) for differential expression tests. Default = 1 (i.e. 2-fold). See details</p></td>
    </tr>
    <tr>
      <th>q1.th</th>
      <td><p>Lower boundary threshold for foreground detection in proportional comparisons. Default = 0.5. See details.</p></td>
    </tr>
    <tr>
      <th>q2.th</th>
      <td><p>Upper boundary threshold for background detection in proportional comparisons. Default = NULL. See details.</p></td>
    </tr>
    <tr>
      <th>q.diff.th</th>
      <td><p>Threshold for scaled difference in proportions. Default = 0.7. See details.</p></td>
    </tr>
    <tr>
      <th>de.score.th</th>
      <td><p>Lower boundary of total differential expression scores for cluster comparisons. Default = 150. See details.</p></td>
    </tr>
    <tr>
      <th>min.cells</th>
      <td><p>The minimum number of cells allowed in each cluster. Default = 4. See details.</p></td>
    </tr>
    <tr>
      <th>min.genes</th>
      <td><p>The minimum number of differentially expressed genes required to separate clusters. Default = 5. See details.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>returns a list of parameters for reuse</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Calling <code>de.param()</code> without additional parameters provides reasonable defaults for high depth (e.g. SMART-seq) datasets.</p>
    <p><strong>Gene detection threshold:</strong></p>
<p><code>low.th</code> sets a lower bound for normalized gene expression to determine whether or not a gene is considered to be detected.
This is used to filter genes that are too low in expression to be reliably detected.<br />
This parameter can be set globally by providing a single value, or per-gene by providing a named vector.</p>
<p><strong>Differential expression test thresholds:</strong></p>
<p><code>scrattch.hicat</code> utilizes <code>limma</code>'s <a href='https://www.rdocumentation.org/packages/limma/topics/ebayes'>eBayes</a> or Chi-Square tests for differential gene expression. These parameters are used to 
determine which genes are considered differentially expressed:<br />
<code>padj.th</code> is the threshold for adjusted p-values. Adjusted p-values must be below this threshold to be considered significant.<br /> 
<code>lfc.th</code> is the threshold for abs(log2(Fold Change)).</p>
<p><strong>Cluster proportion thresholds:</strong></p>
<p>We use <code>q1.th</code>, <code>q2.th</code> and <code>q.diff.th</code> for additional tests based on the proportion of cells in each cluster that express each gene.
For every pair of clusters, we define <em>q1</em> and <em>q2</em> as the proportion of cells with expression greater than <code>low.th</code> 
(above) in the foregound and background cluster, respectively. 
We use <code>q1.th</code> to select genes in a high proportion of foreground clusters, and <code>q2.th</code> to select genes in a low proportion of background clusters. 
Finally, we use <code>q.diff.th</code> to test for the difference between the foreground and background proportions.</p>
<p><code>q1.th</code>: The minimum proportion of cells in the foreground cluster with expression greater than <code>low.th</code>.<br />
<code>q2.th</code>: The maximum proportion of cells in the background cluster with expression greater than <code>low.th</code>.<br />
<code>q.diff.th</code>: The scaled proportional difference between <em>q1</em> and <em>q2</em>, defined as \(abs(q1 - q2) / max(q1, q2)\) .</p>
<p><strong>Cluster-wise p-value threshold:</strong></p>
<p>After performing differential expression tests between a pair of clusters, we use <em>de.score</em> as a way to determine if enough overall 
differential expression is observed to consider the two clusters distinct from each other.</p>
<p>We define <em>de.score</em> for each gene as \(min(-log10(p.adj), 20)\). This sets a cap on the contribution of each gene to the cluster-wise <em>de.score</em> value at 20.<br />
The <em>de.score</em> for a pair of clusters is the sum of the gene-wise <em>de.score</em> values.</p>
<p>Only genes passing the <code>padj.th</code> and <code>lfc.th</code> thresholds (above) contribute to the <em>de.score</em>.</p>
<p><code>de.score.th</code> is used as a minimum value for the cluster-wise <em>de.score</em> in a pairwise comparison between clusters.</p>
<p><strong>Cell and gene count thresholds:</strong></p>
<p><code>min.cells</code> is the minimum size allowed for a cluster. If a cluster size is below <code>min.cells</code>, it will be merged with the nearest cluster.<br /></p>
<p><code>min.genes</code> is the minimum number of differentially expressed genes (passing the <code>padj.th</code> and <code>lfc.th</code> thresholds, above) 
required to consider two clusters separate.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'># Recommended initial parameters for SMART-Seq (&gt; 8,000 genes per sample):</span>

<span class='no'>sm_param</span> <span class='kw'>&lt;-</span> <span class='fu'>de_param</span>(<span class='kw'>low.th</span> <span class='kw'>=</span> <span class='fl'>1</span>,
                     <span class='kw'>padj.th</span> <span class='kw'>=</span> <span class='fl'>0.01</span>,
                     <span class='kw'>lfc.th</span> <span class='kw'>=</span> <span class='fl'>1</span>,
                     <span class='kw'>q1.th</span> <span class='kw'>=</span> <span class='fl'>0.5</span>,
                     <span class='kw'>q2.th</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
                     <span class='kw'>q.diff.th</span> <span class='kw'>=</span> <span class='fl'>0.7</span>,
                     <span class='kw'>de.score.th</span> <span class='kw'>=</span> <span class='fl'>150</span>,
                     <span class='kw'>min.cells</span> <span class='kw'>=</span> <span class='fl'>4</span>,
                     <span class='kw'>min.genes</span> <span class='kw'>=</span> <span class='fl'>5</span>)

<span class='co'># Recommended initial parameters for 10x Cells (&gt; 3,000 genes per sample):</span>

<span class='no'>tx_param</span> <span class='kw'>&lt;-</span> <span class='fu'>de_param</span>(<span class='kw'>low.th</span> <span class='kw'>=</span> <span class='fl'>1</span>,
                     <span class='kw'>padj.th</span> <span class='kw'>=</span> <span class='fl'>0.01</span>,
                     <span class='kw'>lfc.th</span> <span class='kw'>=</span> <span class='fl'>1</span>,
                     <span class='kw'>q1.th</span> <span class='kw'>=</span> <span class='fl'>0.4</span>, <span class='co'># Reduced due to dropout</span>
                     <span class='kw'>q2.th</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
                     <span class='kw'>q.diff.th</span> <span class='kw'>=</span> <span class='fl'>0.7</span>,
                     <span class='kw'>de.score.th</span> <span class='kw'>=</span> <span class='fl'>150</span>,
                     <span class='kw'>min.cells</span> <span class='kw'>=</span> <span class='fl'>10</span>, <span class='co'># Increased due to higher number of cells</span>
                     <span class='kw'>min.genes</span> <span class='kw'>=</span> <span class='fl'>5</span>)

<span class='co'># Recommended initial parameters for 10x Nuclei (&gt; 1,000 genes per sample):</span>

<span class='no'>tx_param</span> <span class='kw'>&lt;-</span> <span class='fu'>de_param</span>(<span class='kw'>low.th</span> <span class='kw'>=</span> <span class='fl'>1</span>,
                     <span class='kw'>padj.th</span> <span class='kw'>=</span> <span class='fl'>0.01</span>,
                     <span class='kw'>lfc.th</span> <span class='kw'>=</span> <span class='fl'>1</span>,
                     <span class='kw'>q1.th</span> <span class='kw'>=</span> <span class='fl'>0.3</span>, <span class='co'># Reduced due to dropout</span>
                     <span class='kw'>q2.th</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
                     <span class='kw'>q.diff.th</span> <span class='kw'>=</span> <span class='fl'>0.7</span>,
                     <span class='kw'>de.score.th</span> <span class='kw'>=</span> <span class='fl'>100</span>, <span class='co'># Reduced due to decreased detection</span>
                     <span class='kw'>min.cells</span> <span class='kw'>=</span> <span class='fl'>10</span>, <span class='co'># Increased due to higher number of cells</span>
                     <span class='kw'>min.genes</span> <span class='kw'>=</span> <span class='fl'>5</span>)</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zizhen Yao, Lucas Graybuck, Trygve Bakken, Jeremy Miller, Cindy van Velthoven.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

