<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Set differential expression (DE) parameters for genes and clusters. — de_param • scrattch.hicat</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Set differential expression (DE) parameters for genes and clusters. — de_param"><meta property="og:description" content="This function provides a convenient way to manage settings for differential expression tests in scrattch.hicat."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scrattch.hicat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/scrattch.hicat_release.html">Overview of scrattch hicat</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/AllenInstitute/scrattch.hicat/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Set differential expression (DE) parameters for genes and clusters.</h1>
    <small class="dont-index">Source: <a href="https://github.com/AllenInstitute/scrattch.hicat/blob/HEAD/R/de.genes.R" class="external-link"><code>R/de.genes.R</code></a></small>
    <div class="hidden name"><code>de_param.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function provides a convenient way to manage settings for differential expression tests in scrattch.hicat.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">de_param</span><span class="op">(</span></span>
<span>  low.th <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  padj.th <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  lfc.th <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  q1.th <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  q2.th <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  q.diff.th <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  de.score.th <span class="op">=</span> <span class="fl">150</span>,</span>
<span>  min.cells <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  min.genes <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-low-th">low.th<a class="anchor" aria-label="anchor" href="#arg-low-th"></a></dt>
<dd><p>Lower boundary for normalized gene expression. Default = 1. See details.</p></dd>


<dt id="arg-padj-th">padj.th<a class="anchor" aria-label="anchor" href="#arg-padj-th"></a></dt>
<dd><p>Upper boundary threshold for adjusted p-values for differential expression tests. Default = 0.01. See details.</p></dd>


<dt id="arg-lfc-th">lfc.th<a class="anchor" aria-label="anchor" href="#arg-lfc-th"></a></dt>
<dd><p>Lower boundary threshold for log2(fold change) for differential expression tests. Default = 1 (i.e. 2-fold). See details</p></dd>


<dt id="arg-q-th">q1.th<a class="anchor" aria-label="anchor" href="#arg-q-th"></a></dt>
<dd><p>Lower boundary threshold for foreground detection in proportional comparisons. Default = 0.5. See details.</p></dd>


<dt id="arg-q-th">q2.th<a class="anchor" aria-label="anchor" href="#arg-q-th"></a></dt>
<dd><p>Upper boundary threshold for background detection in proportional comparisons. Default = NULL. See details.</p></dd>


<dt id="arg-q-diff-th">q.diff.th<a class="anchor" aria-label="anchor" href="#arg-q-diff-th"></a></dt>
<dd><p>Threshold for scaled difference in proportions. Default = 0.7. See details.</p></dd>


<dt id="arg-de-score-th">de.score.th<a class="anchor" aria-label="anchor" href="#arg-de-score-th"></a></dt>
<dd><p>Lower boundary of total differential expression scores for cluster comparisons. Default = 150. See details.</p></dd>


<dt id="arg-min-cells">min.cells<a class="anchor" aria-label="anchor" href="#arg-min-cells"></a></dt>
<dd><p>The minimum number of cells allowed in each cluster. Default = 4. See details.</p></dd>


<dt id="arg-min-genes">min.genes<a class="anchor" aria-label="anchor" href="#arg-min-genes"></a></dt>
<dd><p>The minimum number of differentially expressed genes required to separate clusters. Default = 5. See details.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>returns a list of parameters for reuse</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Calling <code>de.param()</code> without additional parameters provides reasonable defaults for high depth (e.g. SMART-seq) datasets.</p>

<p><strong>Gene detection threshold:</strong></p>
<p><code>low.th</code> sets a lower bound for normalized gene expression to determine whether or not a gene is considered to be detected.
This is used to filter genes that are too low in expression to be reliably detected.<br>
This parameter can be set globally by providing a single value, or per-gene by providing a named vector.</p>
<p><strong>Differential expression test thresholds:</strong></p>
<p><code>scrattch.hicat</code> utilizes <code>limma</code>'s <a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a> or Chi-Square tests for differential gene expression. These parameters are used to
determine which genes are considered differentially expressed:<br><code>padj.th</code> is the threshold for adjusted p-values. Adjusted p-values must be below this threshold to be considered significant.<br><code>lfc.th</code> is the threshold for abs(log2(Fold Change)).</p>
<p><strong>Cluster proportion thresholds:</strong></p>
<p>We use <code>q1.th</code>, <code>q2.th</code> and <code>q.diff.th</code> for additional tests based on the proportion of cells in each cluster that express each gene.
For every pair of clusters, we define <em>q1</em> and <em>q2</em> as the proportion of cells with expression greater than <code>low.th</code>
(above) in the foregound and background cluster, respectively.
We use <code>q1.th</code> to select genes in a high proportion of foreground clusters, and <code>q2.th</code> to select genes in a low proportion of background clusters.
Finally, we use <code>q.diff.th</code> to test for the difference between the foreground and background proportions.</p>
<p><code>q1.th</code>: The minimum proportion of cells in the foreground cluster with expression greater than <code>low.th</code>.<br><code>q2.th</code>: The maximum proportion of cells in the background cluster with expression greater than <code>low.th</code>.<br><code>q.diff.th</code>: The scaled proportional difference between <em>q1</em> and <em>q2</em>, defined as \(abs(q1 - q2) / max(q1, q2)\) .</p>
<p><strong>Cluster-wise p-value threshold:</strong></p>
<p>After performing differential expression tests between a pair of clusters, we use <em>de.score</em> as a way to determine if enough overall
differential expression is observed to consider the two clusters distinct from each other.</p>
<p>We define <em>de.score</em> for each gene as \(min(-log10(p.adj), 20)\). This sets a cap on the contribution of each gene to the cluster-wise <em>de.score</em> value at 20.<br>
The <em>de.score</em> for a pair of clusters is the sum of the gene-wise <em>de.score</em> values.</p>
<p>Only genes passing the <code>padj.th</code> and <code>lfc.th</code> thresholds (above) contribute to the <em>de.score</em>.</p>
<p><code>de.score.th</code> is used as a minimum value for the cluster-wise <em>de.score</em> in a pairwise comparison between clusters.</p>
<p><strong>Cell and gene count thresholds:</strong></p>
<p><code>min.cells</code> is the minimum size allowed for a cluster. If a cluster size is below <code>min.cells</code>, it will be merged with the nearest cluster.<br></p>
<p><code>min.genes</code> is the minimum number of differentially expressed genes (passing the <code>padj.th</code> and <code>lfc.th</code> thresholds, above)
required to consider two clusters separate.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span></span>
<span><span class="co"># Recommended initial parameters for SMART-Seq (&gt; 8,000 genes per sample):</span></span>
<span></span>
<span><span class="va">sm_param</span> <span class="op">&lt;-</span> <span class="fu">de_param</span><span class="op">(</span>low.th <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                     padj.th <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>                     lfc.th <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                     q1.th <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                     q2.th <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                     q.diff.th <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>                     de.score.th <span class="op">=</span> <span class="fl">150</span>,</span>
<span>                     min.cells <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                     min.genes <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Recommended initial parameters for 10x Cells (&gt; 3,000 genes per sample):</span></span>
<span></span>
<span><span class="va">tx_param</span> <span class="op">&lt;-</span> <span class="fu">de_param</span><span class="op">(</span>low.th <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                     padj.th <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>                     lfc.th <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                     q1.th <span class="op">=</span> <span class="fl">0.4</span>, <span class="co"># Reduced due to dropout</span></span>
<span>                     q2.th <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                     q.diff.th <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>                     de.score.th <span class="op">=</span> <span class="fl">150</span>,</span>
<span>                     min.cells <span class="op">=</span> <span class="fl">10</span>, <span class="co"># Increased due to higher number of cells</span></span>
<span>                     min.genes <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Recommended initial parameters for 10x Nuclei (&gt; 1,000 genes per sample):</span></span>
<span></span>
<span><span class="va">tx_param</span> <span class="op">&lt;-</span> <span class="fu">de_param</span><span class="op">(</span>low.th <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                     padj.th <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>                     lfc.th <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                     q1.th <span class="op">=</span> <span class="fl">0.3</span>, <span class="co"># Reduced due to dropout</span></span>
<span>                     q2.th <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                     q.diff.th <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>                     de.score.th <span class="op">=</span> <span class="fl">100</span>, <span class="co"># Reduced due to decreased detection</span></span>
<span>                     min.cells <span class="op">=</span> <span class="fl">10</span>, <span class="co"># Increased due to higher number of cells</span></span>
<span>                     min.genes <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> </span>
<span></span></code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Zizhen Yao, Lucas Graybuck, Trygve Bakken, Cindy van Velthoven, Jeremy Miller, Adriana Sedeno-Cortes, Changkyu Lee, Lawrence Huang, Saroja Somasundaram.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

