<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Package index â€¢ scrattch.hicat</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scrattch.hicat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/scrattch.hicat_release.html">Overview of scrattch hicat</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/AllenInstitute/scrattch.hicat/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="CCA.html">CCA()</a></code> </p>
        </td>
        <td><p>CCA</p></td>
      </tr><tr><td>
          <p><code><a href="CanonCor.html">CanonCor()</a></code> </p>
        </td>
        <td><p>Canon cor</p></td>
      </tr><tr><td>
          <p><code><a href="DE_genes_cat_by_cl.html">DE_genes_cat_by_cl()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="DE_genes_pairs.html">DE_genes_pairs()</a></code> </p>
        </td>
        <td><p>reverse-compatibility wrapper around de_selected_pairs()</p></td>
      </tr><tr><td>
          <p><code><a href="DE_genes_pw.html">DE_genes_pw()</a></code> </p>
        </td>
        <td><p>reverse-compatibility wrapper around de_all_pairs()</p></td>
      </tr><tr><td>
          <p><code><a href="RunUMAP.html">RunUMAP(<i>&lt;default&gt;</i>)</a></code> </p>
        </td>
        <td><p>Adapt from Seurat package</p></td>
      </tr><tr><td>
          <p><code><a href="adjust_color.html">adjust_color()</a></code> </p>
        </td>
        <td><p>Correct duplicated colors</p></td>
      </tr><tr><td>
          <p><code><a href="angle.html">angle()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="avgangle.html">avgangle()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="batch_process.html">batch_process()</a></code> </p>
        </td>
        <td><p>Batch process</p></td>
      </tr><tr><td>
          <p><code><a href="build_dend.html">build_dend()</a></code> </p>
        </td>
        <td><p>Build dend</p></td>
      </tr><tr><td>
          <p><code><a href="build_dend_harmonize.html">build_dend_harmonize()</a></code> </p>
        </td>
        <td><p>Build dend harmonize</p></td>
      </tr><tr><td>
          <p><code><a href="build_dend_with_means.html">build_dend_with_means()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="build_tsne.html">build_tsne()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="calc_tau.html">calc_tau()</a></code> </p>
        </td>
        <td><p>Calculate Tau scores for each gene</p></td>
      </tr><tr><td>
          <p><code><a href="check_donor.html">check_donor()</a></code> </p>
        </td>
        <td><p>Find donor clusters</p></td>
      </tr><tr><td>
          <p><code><a href="check_neun.html">check_neun()</a></code> </p>
        </td>
        <td><p>Check whether clusters are Neun+ or Neun-</p></td>
      </tr><tr><td>
          <p><code><a href="check_outlier.html">check_outlier()</a></code> </p>
        </td>
        <td><p>Check for outlier clusters</p></td>
      </tr><tr><td>
          <p><code><a href="check_qc.html">check_qc()</a></code> </p>
        </td>
        <td><p>Check qc of numeric metric</p></td>
      </tr><tr><td>
          <p><code><a href="clean_outliers.html">clean_outliers()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="collect_co_matrix.html">collect_co_matrix()</a></code> </p>
        </td>
        <td><p>Collect co-clustering matrix from results files</p></td>
      </tr><tr><td>
          <p><code><a href="comb_de_result.html">comb_de_result()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="combine_cl.html">combine_cl()</a></code> </p>
        </td>
        <td><p>combine cl</p></td>
      </tr><tr><td>
          <p><code><a href="compare_annotate.html">compare_annotate()</a></code> </p>
        </td>
        <td><p>Compare two sets of cluster assignments for the same set of cells</p></td>
      </tr><tr><td>
          <p><code><a href="compare_plot.html">compare_plot()</a></code> </p>
        </td>
        <td><p>Compare and plot two sets of cluster assignments</p></td>
      </tr><tr><td>
          <p><code><a href="compute_knn.html">compute_knn()</a></code> </p>
        </td>
        <td><p>compute knn</p></td>
      </tr><tr><td>
          <p><code><a href="compute_vg_stats.html">compute_vg_stats()</a></code> </p>
        </td>
        <td><p>Compute gene expression variance statistics</p></td>
      </tr><tr><td>
          <p><code><a href="convert_pair_matrix.html">convert_pair_matrix()</a></code> </p>
        </td>
        <td><p>Convert paired cluster comparison values to a matrix</p></td>
      </tr><tr><td>
          <p><code><a href="convert_pair_matrix_str.html">convert_pair_matrix_str()</a></code> </p>
        </td>
        <td><p>Convert paired cluster comparison values to a matrix</p></td>
      </tr><tr><td>
          <p><code><a href="cpm.html">cpm()</a></code> </p>
        </td>
        <td><p>Convert a matrix of raw counts to a matrix of Counts per Million values</p></td>
      </tr><tr><td>
          <p><code><a href="create_pairs.html">create_pairs()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="cutree_dend.html">cutree_dend()</a></code> </p>
        </td>
        <td><p>Cut tree dendrogram</p></td>
      </tr><tr><td>
          <p><code><a href="de_all_pairs.html">de_all_pairs()</a></code> </p>
        </td>
        <td><p>Perform all pairwise differential expression comparison between clusters</p></td>
      </tr><tr><td>
          <p><code><a href="de_genes_pairs_multiple.html">de_genes_pairs_multiple()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="de_pair.html">de_pair()</a></code> </p>
        </td>
        <td><p>reverse-compatibility wrapper around score_pair_stats()</p></td>
      </tr><tr><td>
          <p><code><a href="de_pair_chisq.html">de_pair_chisq()</a></code> </p>
        </td>
        <td><p>Perform pairwise differential detection tests using Chi-Squared for a single pair of clusters</p></td>
      </tr><tr><td>
          <p><code><a href="de_pair_limma.html">de_pair_limma()</a></code> </p>
        </td>
        <td><p>Perform pairwise DE tests using limma for a single pair of clusters</p></td>
      </tr><tr><td>
          <p><code><a href="de_param.html">de_param()</a></code> </p>
        </td>
        <td><p>Set differential expression (DE) parameters for genes and clusters.</p></td>
      </tr><tr><td>
          <p><code><a href="de_score.html">de_score()</a></code> </p>
        </td>
        <td><p>reverse-compatibility wrapper around de_all_pairs_stats()</p></td>
      </tr><tr><td>
          <p><code><a href="de_score_pairs.html">de_score_pairs()</a></code> </p>
        </td>
        <td><p>reverse-compatibility wrapper around de_selected_pairs_stats()</p></td>
      </tr><tr><td>
          <p><code><a href="de_selected_pairs.html">de_selected_pairs()</a></code> </p>
        </td>
        <td><p>Perform pairwise differential gene expression tests between main pairs of clusters in parallel</p></td>
      </tr><tr><td>
          <p><code><a href="de_stats_all_pairs.html">de_stats_all_pairs()</a></code> </p>
        </td>
        <td><p>Compute differential expression summary statistics for all pairs of clusters based on de_param()</p></td>
      </tr><tr><td>
          <p><code><a href="de_stats_pair.html">de_stats_pair()</a></code> </p>
        </td>
        <td><p>Compute differential expression summary statistics based on a differential results data.frame and de_param().</p></td>
      </tr><tr><td>
          <p><code><a href="de_stats_selected_pairs.html">de_stats_selected_pairs()</a></code> </p>
        </td>
        <td><p>Compute differential expression summary statistics for selected pairs of clusters based on de_param().</p></td>
      </tr><tr><td>
          <p><code><a href="dend_lca.html">dend_lca()</a></code> </p>
        </td>
        <td><p>Dend lca</p></td>
      </tr><tr><td>
          <p><code><a href="dend_list.html">dend_list()</a></code> </p>
        </td>
        <td><p>Dendrogram list</p></td>
      </tr><tr><td>
          <p><code><a href="dend_match.html">dend_match()</a></code> </p>
        </td>
        <td><p>Dendrogram match</p></td>
      </tr><tr><td>
          <p><code><a href="display_cl.html">display_cl()</a></code> </p>
        </td>
        <td><p>Display cluster plot</p></td>
      </tr><tr><td>
          <p><code><a href="display_cl4.html">display_cl4()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="display_cl_markers_co.ratio.html">display_cl_markers_co.ratio()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="display_cl_one_vs_others.html">display_cl_one_vs_others()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="doubletFinder.html">doubletFinder()</a></code> </p>
        </td>
        <td><p>Doublet detection in single-cell RNA sequencing data</p></td>
      </tr><tr><td>
          <p><code><a href="edgeMaker.html">edgeMaker()</a></code> </p>
        </td>
        <td><p>function to draw (curved) line between two points</p></td>
      </tr><tr><td>
          <p><code><a href="fast_tsne.html">fast_tsne()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="findVG.html">findVG()</a></code> </p>
        </td>
        <td><p>reverse-compatibility wrapper around find_vg()</p></td>
      </tr><tr><td>
          <p><code><a href="find_doublet.html">find_doublet()</a></code> </p>
        </td>
        <td><p>Find doublet</p></td>
      </tr><tr><td>
          <p><code><a href="find_low_quality_cl.html">find_low_quality_cl()</a></code> </p>
        </td>
        <td><p>Use DEGenes to identify clusters with few differentially expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="find_vg.html">find_vg()</a></code> </p>
        </td>
        <td><p>Identify high variability genes using Brennecke's method</p></td>
      </tr><tr><td>
          <p><code><a href="gene_dispersion.html">gene_dispersion()</a></code> </p>
        </td>
        <td><p>Compute dispersion for each gene in a gene x sample matrix</p></td>
      </tr><tr><td>
          <p><code><a href="gene_gene_cor_conservation.html">gene_gene_cor_conservation()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="gene_loess_fit.html">gene_loess_fit()</a></code> </p>
        </td>
        <td><p>Compute dispersion-mean Loess-fit for each gene in a gene x sample matrix</p></td>
      </tr><tr><td>
          <p><code><a href="gene_loess_fit_z.html">gene_loess_fit_z()</a></code> </p>
        </td>
        <td><p>Compute dispersion z scores for the residuals of a loess fit</p></td>
      </tr><tr><td>
          <p><code><a href="gene_means.html">gene_means()</a></code> </p>
        </td>
        <td><p>Compute means for each gene in a gene x sample matrix</p></td>
      </tr><tr><td>
          <p><code><a href="gene_vars.html">gene_vars()</a></code> </p>
        </td>
        <td><p>Compute variance for each gene in a gene x sample matrix</p></td>
      </tr><tr><td>
          <p><code><a href="gene_z.html">gene_z()</a></code> </p>
        </td>
        <td><p>Compute dispersion z-scores for each gene in a gene x sample matrix</p></td>
      </tr><tr><td>
          <p><code><a href="get.markers.num.html">get.markers.num()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="getTopMarkersByPropNew.html">getTopMarkersByPropNew()</a></code> </p>
        </td>
        <td><p>Find specific marker gene for each cluster</p></td>
      </tr><tr><td>
          <p><code><a href="get_RD_cl_center.html">get_RD_cl_center()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="get_beta_score.html">get_beta_score()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="get_cell.cl.co.ratio.html">get_cell.cl.co.ratio()</a></code> </p>
        </td>
        <td><p>Get cell co-clustering ratios</p></td>
      </tr><tr><td>
          <p><code><a href="get_cl_df.html">get_cl_df()</a></code> </p>
        </td>
        <td><p>Generate an initial cl.df object based on cl</p></td>
      </tr><tr><td>
          <p><code><a href="get_cl_mat.html">get_cl_mat()</a></code> </p>
        </td>
        <td><p>Generate a sparse matrix one-hot representation of clusters x samples</p></td>
      </tr><tr><td>
          <p><code><a href="get_cl_means.html">get_cl_means()</a></code> </p>
        </td>
        <td><p>Compute cluster means for each row in a matrix</p></td>
      </tr><tr><td>
          <p><code><a href="get_cl_means_list.html">get_cl_means_list()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="get_cl_medians.html">get_cl_medians()</a></code> </p>
        </td>
        <td><p>Compute cluster medians for each row in a matrix</p></td>
      </tr><tr><td>
          <p><code><a href="get_cl_present_list.html">get_cl_present_list()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="get_cl_prop.html">get_cl_prop()</a></code> </p>
        </td>
        <td><p>Compute cluster proportions for each row in a matrix</p></td>
      </tr><tr><td>
          <p><code><a href="get_cl_sim.html">get_cl_sim()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="get_cl_sim_multiple.html">get_cl_sim_multiple()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="get_cl_sums.html">get_cl_sums()</a></code> </p>
        </td>
        <td><p>Compute cluster sums for each row in a matrix</p></td>
      </tr><tr><td>
          <p><code><a href="get_core_intermediate.html">get_core_intermediate()</a></code> </p>
        </td>
        <td><p>Get core intermediate</p></td>
      </tr><tr><td>
          <p><code><a href="get_de_genes_sym.html">get_de_genes_sym()</a></code> </p>
        </td>
        <td><p>Get de genes sym</p></td>
      </tr><tr><td>
          <p><code><a href="get_de_lfc_list.html">get_de_lfc_list()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="get_de_matrix.html">get_de_matrix()</a></code> </p>
        </td>
        <td><p>Generate a matrix of pairwise DE results</p></td>
      </tr><tr><td>
          <p><code><a href="get_de_pair.html">get_de_pair()</a></code> </p>
        </td>
        <td><p>Get de pair</p></td>
      </tr><tr><td>
          <p><code><a href="get_de_result.html">get_de_result()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="get_de_score.html">get_de_score()</a></code> </p>
        </td>
        <td><p>Get de score</p></td>
      </tr><tr><td>
          <p><code><a href="get_dend_markers.html">get_dend_markers()</a></code> </p>
        </td>
        <td><p>get dendrogram precomputed markers</p></td>
      </tr><tr><td>
          <p><code><a href="get_dend_markers_direction.html">get_dend_markers_direction()</a></code> </p>
        </td>
        <td><p>get dendrogram precomputed markers (direction)</p></td>
      </tr><tr><td>
          <p><code><a href="get_dend_parent.html">get_dend_parent()</a></code> </p>
        </td>
        <td><p>Get dendrogram parent</p></td>
      </tr><tr><td>
          <p><code><a href="get_eigen.html">get_eigen()</a></code> </p>
        </td>
        <td><p>Compute module eigen genes</p></td>
      </tr><tr><td>
          <p><code><a href="get_gene_cl_correlation.html">get_gene_cl_correlation()</a></code> </p>
        </td>
        <td><p>Get gene cl correlation</p></td>
      </tr><tr><td>
          <p><code><a href="get_gene_score.html">get_gene_score()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="get_knn.html">get_knn()</a></code> </p>
        </td>
        <td><p>Get KNN</p></td>
      </tr><tr><td>
          <p><code><a href="get_knn_batch.html">get_knn_batch()</a></code> </p>
        </td>
        <td><p>get knn batch</p></td>
      </tr><tr><td>
          <p><code><a href="get_knn_graph.html">get_knn_graph()</a></code> </p>
        </td>
        <td><p>get knn graph</p></td>
      </tr><tr><td>
          <p><code><a href="get_knn_weight.html">get_knn_weight()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="get_pair_matrix.html">get_pair_matrix()</a></code> </p>
        </td>
        <td><p>Subset a matrix as a vector using row and column positions</p></td>
      </tr><tr><td>
          <p><code><a href="get_pair_matrix_coor.html">get_pair_matrix_coor()</a></code> </p>
        </td>
        <td><p>Convert matrix row/column positions to vector position</p></td>
      </tr><tr><td>
          <p><code><a href="get_pairs.html">get_pairs()</a></code> </p>
        </td>
        <td><p>Convert underscore_separated pair names to a data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="group_cl.html">group_cl()</a></code> </p>
        </td>
        <td><p>Assign clusters to a group</p></td>
      </tr><tr><td>
          <p><code><a href="group_specific_markers.html">group_specific_markers()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="harmonize.html">harmonize()</a></code> </p>
        </td>
        <td><p>Harmonize</p></td>
      </tr><tr><td>
          <p><code><a href="heatmap.3.html">heatmap.3()</a></code> </p>
        </td>
        <td><p>A modified version of heatmap.2 from the gplots package for cluster plotting</p></td>
      </tr><tr><td>
          <p><code><a href="heatmap.4.html">heatmap.4()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="i_harmonize.html">i_harmonize()</a></code> </p>
        </td>
        <td><p>i harmonize</p></td>
      </tr><tr><td>
          <p><code><a href="impute_knn.html">impute_knn()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="impute_knn_global.html">impute_knn_global()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="impute_val_cor.html">impute_val_cor()</a></code> </p>
        </td>
        <td><p>Impute val cor</p></td>
      </tr><tr><td>
          <p><code><a href="incrementHex.html">incrementHex()</a></code> </p>
        </td>
        <td><p>Adjust color vector</p></td>
      </tr><tr><td>
          <p><code><a href="iter_clust.html">iter_clust()</a></code> </p>
        </td>
        <td><p>Iterative clustering algorithm for single cell RNAseq dataset</p></td>
      </tr><tr><td>
          <p><code><a href="iter_consensus_clust.html">iter_consensus_clust()</a></code> </p>
        </td>
        <td><p>Iterative consensus clustering</p></td>
      </tr><tr><td>
          <p><code><a href="iter_impute_knn.html">iter_impute_knn()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="jaccard.html">jaccard()</a></code> </p>
        </td>
        <td><p>Compute jaccard distances between columns of a matrix</p></td>
      </tr><tr><td>
          <p><code><a href="jaccard_leiden.RANN.html">jaccard_leiden.RANN()</a></code> </p>
        </td>
        <td><p>Perform Jaccard/Louvain clustering based on RANN</p></td>
      </tr><tr><td>
          <p><code><a href="jaccard_leiden.html">jaccard_leiden()</a></code> </p>
        </td>
        <td><p>Perform Jaccard/Leiden clustering</p></td>
      </tr><tr><td>
          <p><code><a href="jaccard_louvain.RANN.html">jaccard_louvain.RANN()</a></code> </p>
        </td>
        <td><p>Perform Jaccard/Louvain clustering based on RANN</p></td>
      </tr><tr><td>
          <p><code><a href="jaccard_louvain.html">jaccard_louvain()</a></code> </p>
        </td>
        <td><p>Perform Jaccard/Louvain clustering using Rphenograph</p></td>
      </tr><tr><td>
          <p><code><a href="knn_cor.html">knn_cor()</a></code> </p>
        </td>
        <td><p>KNN cor</p></td>
      </tr><tr><td>
          <p><code><a href="knn_cosine.html">knn_cosine()</a></code> </p>
        </td>
        <td><p>KNN cosine</p></td>
      </tr><tr><td>
          <p><code><a href="knn_jaccard_louvain.html">knn_jaccard_louvain()</a></code> </p>
        </td>
        <td><p>KNN Jaccard Louvain</p></td>
      </tr><tr><td>
          <p><code><a href="knn_joint.html">knn_joint()</a></code> </p>
        </td>
        <td><p>knn joint</p></td>
      </tr><tr><td>
          <p><code><a href="label_dend.html">label_dend()</a></code> </p>
        </td>
        <td><p>Label dendrogram</p></td>
      </tr><tr><td>
          <p><code><a href="lm_matrix.html">lm_matrix()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="lm_normalize.html">lm_normalize()</a></code> </p>
        </td>
        <td><p>add batch substracted median</p></td>
      </tr><tr><td>
          <p><code><a href="logCPM.html">logCPM()</a></code> </p>
        </td>
        <td><p>Convert a matrix of raw counts to a matrix of log2(Counts per Million + 1) values</p></td>
      </tr><tr><td>
          <p><code><a href="makeColorsUnique.html">makeColorsUnique()</a></code> </p>
        </td>
        <td><p>Make a vector of unique colors</p></td>
      </tr><tr><td>
          <p><code><a href="makeLayerLabel.html">makeLayerLabel()</a></code> </p>
        </td>
        <td><p>Numeric layer for annotation object</p></td>
      </tr><tr><td>
          <p><code><a href="makeRegionLabel.html">makeRegionLabel()</a></code> </p>
        </td>
        <td><p>Split region from roi for annotation object</p></td>
      </tr><tr><td>
          <p><code><a href="map_by_cor.html">map_by_cor()</a></code> </p>
        </td>
        <td><p>Map samples to a training dataset by correlation</p></td>
      </tr><tr><td>
          <p><code><a href="map_cl_summary.html">map_cl_summary()</a></code> </p>
        </td>
        <td><p>Map a dataset to a reference, and compare existing cluster calls to the reference comparison</p></td>
      </tr><tr><td>
          <p><code><a href="map_cv.html">map_cv()</a></code> </p>
        </td>
        <td><p>Run a single round of cross-validation of cluster mapping using a subset of marker genes</p></td>
      </tr><tr><td>
          <p><code><a href="map_dend.html">map_dend()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="map_dend_markers.html">map_dend_markers()</a></code> </p>
        </td>
        <td><p>Map dend markers</p></td>
      </tr><tr><td>
          <p><code><a href="map_dend_membership.html">map_dend_membership()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="map_sampling.html">map_sampling()</a></code> </p>
        </td>
        <td><p>Perform bootstrapped mapping using a fraction of provided marker genes.</p></td>
      </tr><tr><td>
          <p><code><a href="mapping.html">mapping()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="markers_max_tau.html">markers_max_tau()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="markers_tau_one_vs_other.html">markers_tau_one_vs_other()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="match_cl.html">match_cl()</a></code> </p>
        </td>
        <td><p>Compute correlations of clusters to a reference set, and get the best-correlated reference clusters</p></td>
      </tr><tr><td>
          <p><code><a href="merge_cl.html">merge_cl()</a></code> </p>
        </td>
        <td><p>Merge clusters based on pairwise differential expressed genes.</p></td>
      </tr><tr><td>
          <p><code><a href="merge_cl_multiple.html">merge_cl_multiple()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="merge_knn_result.html">merge_knn_result()</a></code> </p>
        </td>
        <td><p>merge knn result</p></td>
      </tr><tr><td>
          <p><code><a href="multiplot.html">multiplot()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="mybgplot3d.html">mybgplot3d()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="node_specific_markers.html">node_specific_markers()</a></code> </p>
        </td>
        <td><p>Node specific markers</p></td>
      </tr><tr><td>
          <p><code><a href="node_vs_sibling_markers.html">node_vs_sibling_markers()</a></code> </p>
        </td>
        <td><p>Node vs sibling markers</p></td>
      </tr><tr><td>
          <p><code><a href="onestep_clust.html">onestep_clust()</a></code> </p>
        </td>
        <td><p>One round of clustering in the iteractive clustering pipeline</p></td>
      </tr><tr><td>
          <p><code><a href="pair_cor.html">pair_cor()</a></code> </p>
        </td>
        <td><p>Compute correlations between each matching row or column of two matrices</p></td>
      </tr><tr><td>
          <p><code><a href="perp.html">perp()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="perpEnd.html">perpEnd()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="perpMid.html">perpMid()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="perpStart.html">perpStart()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_2d_select.html">plot_2d_select()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_3d_label.html">plot_3d_label()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_3d_label_multiple.html">plot_3d_label_multiple()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_3d_select.html">plot_3d_select()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_3d_val.html">plot_3d_val()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_RD_cl.html">plot_RD_cl()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_RD_gene.html">plot_RD_gene()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_RD_meta.html">plot_RD_meta()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_cl_cells.html">plot_cl_cells()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_cl_heatmap.html">plot_cl_heatmap()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_cl_heatmap4.html">plot_cl_heatmap4()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_cl_low.html">plot_cl_low()</a></code> </p>
        </td>
        <td><p>Plot cl low</p></td>
      </tr><tr><td>
          <p><code><a href="plot_cl_meta_barplot.html">plot_cl_meta_barplot()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_confusion.html">plot_confusion()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_de_lfc_num.html">plot_de_lfc_num()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_de_num.html">plot_de_num()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_dend.html">plot_dend()</a></code> </p>
        </td>
        <td><p>Plot dendrogram</p></td>
      </tr><tr><td>
          <p><code><a href="plot_doublet.html">plot_doublet()</a></code> </p>
        </td>
        <td><p>Plot doublet</p></td>
      </tr><tr><td>
          <p><code><a href="plot_low_qc.html">plot_low_qc()</a></code> </p>
        </td>
        <td><p>Plot low-quality clusters</p></td>
      </tr><tr><td>
          <p><code><a href="plot_markers.html">plot_markers()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_markers_cl_means.html">plot_markers_cl_means()</a></code> </p>
        </td>
        <td><p>Plot markers cl means</p></td>
      </tr><tr><td>
          <p><code><a href="plot_pair_matrix.html">plot_pair_matrix()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_tSNE_gene.html">plot_tSNE_gene()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_tSNE_meta.html">plot_tSNE_meta()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_tsne.html">plot_tsne()</a></code> </p>
        </td>
        <td><p>Plot tsne</p></td>
      </tr><tr><td>
          <p><code><a href="plot_tsne_cl.html">plot_tsne_cl()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_vg.html">plot_vg()</a></code> </p>
        </td>
        <td><p>Generate gene variance plots</p></td>
      </tr><tr><td>
          <p><code><a href="prcomp.irlba.html">prcomp(<i>&lt;irlba&gt;</i>)</a></code> </p>
        </td>
        <td><p>prcomp irlba</p></td>
      </tr><tr><td>
          <p><code><a href="predict_annotate_cor.html">predict_annotate_cor()</a></code> </p>
        </td>
        <td><p>Predict annotations by cluster correlation</p></td>
      </tr><tr><td>
          <p><code><a href="predict_knn.html">predict_knn()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="predict_knn_new.html">predict_knn_new()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="prune_dend.html">prune_dend()</a></code> </p>
        </td>
        <td><p>Prune dendrogram</p></td>
      </tr><tr><td>
          <p><code><a href="pvclust_show_signif_gradient.html">pvclust_show_signif_gradient()</a></code> </p>
        </td>
        <td><p>Pv clust show significant gradient</p></td>
      </tr><tr><td>
          <p><code><a href="rd_PCA.html">rd_PCA()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="refine_cl.html">refine_cl()</a></code> </p>
        </td>
        <td><p>Refine clusters</p></td>
      </tr><tr><td>
          <p><code><a href="renameAndOrderClusters.html">renameAndOrderClusters()</a></code> </p>
        </td>
        <td><p>Rename clusters using genes and metadata</p></td>
      </tr><tr><td>
          <p><code><a href="reorder_cl.html">reorder_cl()</a></code> </p>
        </td>
        <td><p>Reorder cluster based on hiearchical clustering of clusters based on average cluster values for the input data matrix</p></td>
      </tr><tr><td>
          <p><code><a href="reorder_dend.html">reorder_dend()</a></code> </p>
        </td>
        <td><p>Reorder dendrogram</p></td>
      </tr><tr><td>
          <p><code><a href="reorder_factor.html">reorder_factor()</a></code> </p>
        </td>
        <td><p>Reorder factors of one annotation to match another</p></td>
      </tr><tr><td>
          <p><code><a href="rescale_samples.html">rescale_samples()</a></code> </p>
        </td>
        <td><p>Rescale a gene x sample matrix</p></td>
      </tr><tr><td>
          <p><code><a href="resolve_cl.html">resolve_cl()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="run_consensus_clust.html">run_consensus_clust()</a></code> </p>
        </td>
        <td><p>Wrapper function to repeatively run clustering on subsampled cells and infer consensus clusters</p></td>
      </tr><tr><td>
          <p><code><a href="sample_cells.html">sample_cells()</a></code> </p>
        </td>
        <td><p>Downsample cells from each cluster</p></td>
      </tr><tr><td>
          <p><code><a href="sample_sets_list.html">sample_sets_list()</a></code> </p>
        </td>
        <td><p>Sample sets lists</p></td>
      </tr><tr><td>
          <p><code><a href="select_N_markers.html">select_N_markers()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="select_dend_markers.html">select_dend_markers()</a></code> </p>
        </td>
        <td><p>Select dend markers</p></td>
      </tr><tr><td>
          <p><code><a href="select_joint_genes.html">select_joint_genes()</a></code> </p>
        </td>
        <td><p>Select joint genes</p></td>
      </tr><tr><td>
          <p><code><a href="select_joint_markers.html">select_joint_markers()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="select_markers.html">select_markers()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="select_markers_pair.html">select_markers_pair()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="select_markers_pair_direction.html">select_markers_pair_direction()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="select_markers_pair_group.html">select_markers_pair_group()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="select_node_specific_markers.html">select_node_specific_markers()</a></code> </p>
        </td>
        <td><p>Select node specific markers</p></td>
      </tr><tr><td>
          <p><code><a href="select_pos_dend_markers.html">select_pos_dend_markers()</a></code> </p>
        </td>
        <td><p>Select pos dend markers</p></td>
      </tr><tr><td>
          <p><code><a href="select_pos_markers.html">select_pos_markers()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="set_pair_matrix.html">set_pair_matrix()</a></code> </p>
        </td>
        <td><p>Update a matrix with values from a 1d vector using row and column positions</p></td>
      </tr><tr><td>
          <p><code><a href="sim_knn.html">sim_knn()</a></code> </p>
        </td>
        <td><p>Sim knn</p></td>
      </tr><tr><td>
          <p><code><a href="simple_dend.html">simple_dend()</a></code> </p>
        </td>
        <td><p>Simple dendrogram</p></td>
      </tr><tr><td>
          <p><code><a href="sparse_cor.html">sparse_cor()</a></code> </p>
        </td>
        <td><p>Compute correlation scores for columns of a sparse matrix</p></td>
      </tr><tr><td>
          <p><code><a href="summarize_cl.html">summarize_cl()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="test_knn.html">test_knn()</a></code> </p>
        </td>
        <td><p>Test knn</p></td>
      </tr><tr><td>
          <p><code><a href="test_merge.html">test_merge()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="unbranch_by_conf.html">unbranch_by_conf()</a></code> </p>
        </td>
        <td><p>Unbranch by conf</p></td>
      </tr><tr><td>
          <p><code><a href="unbranch_by_length.html">unbranch_by_length()</a></code> </p>
        </td>
        <td><p>Unbranch by length</p></td>
      </tr><tr><td>
          <p><code><a href="updateSampDat.html">updateSampDat()</a></code> </p>
        </td>
        <td><p>Update cluster names in annotation data frame</p></td>
      </tr><tr><td>
          <p><code><a href="vec_chisq_test.html">vec_chisq_test()</a></code> </p>
        </td>
        <td><p>Vectorized Chi-squared tests for differential gene detection</p></td>
      </tr><tr><td>
          <p><code><a href="within_group_specific_markers.html">within_group_specific_markers()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Zizhen Yao, Lucas Graybuck, Trygve Bakken, Cindy van Velthoven, Jeremy Miller, Adriana Sedeno-Cortes, Changkyu Lee, Lawrence Huang, Saroja Somasundaram.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

