<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ scrattch.hicat</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Function reference" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scrattch.hicat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.23</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="adjust_color.html">adjust_color()</a></code> </p>
        </td>
        <td><p>Correct duplicated colors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calc_tau.html">calc_tau()</a></code> </p>
        </td>
        <td><p>Calculate Tau scores for each gene</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="check_donor.html">check_donor()</a></code> </p>
        </td>
        <td><p>Find donor clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="check_neun.html">check_neun()</a></code> </p>
        </td>
        <td><p>Check whether clusters are Neun+ or Neun-</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="check_outlier.html">check_outlier()</a></code> </p>
        </td>
        <td><p>Check for outlier clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="check_qc.html">check_qc()</a></code> </p>
        </td>
        <td><p>Check qc of numeric metric</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="collect_co_matrix.html">collect_co_matrix()</a></code> </p>
        </td>
        <td><p>Collect co-clustering matrix from results files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compare_annotate.html">compare_annotate()</a></code> </p>
        </td>
        <td><p>Compare two sets of cluster assignments for the same set of cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compare_plot.html">compare_plot()</a></code> </p>
        </td>
        <td><p>Compare and plot two sets of cluster assignments</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compute_vg_stats.html">compute_vg_stats()</a></code> </p>
        </td>
        <td><p>Compute gene expression variance statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convert_pair_matrix.html">convert_pair_matrix()</a></code> </p>
        </td>
        <td><p>Convert paired cluster comparison values to a matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convert_pair_matrix_str.html">convert_pair_matrix_str()</a></code> </p>
        </td>
        <td><p>Convert paired cluster comparison values to a matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cpm.html">cpm()</a></code> </p>
        </td>
        <td><p>Convert a matrix of raw counts to a matrix of Counts per Million values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="de_all_pairs.html">de_all_pairs()</a></code> </p>
        </td>
        <td><p>Perform all pairwise differential expression comparison between clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="DE_genes_pairs.html">DE_genes_pairs()</a></code> </p>
        </td>
        <td><p>reverse-compatibility wrapper around de_selected_pairs()</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="DE_genes_pw.html">DE_genes_pw()</a></code> </p>
        </td>
        <td><p>reverse-compatibility wrapper around de_all_pairs()</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="de_pair.html">de_pair()</a></code> </p>
        </td>
        <td><p>reverse-compatibility wrapper around score_pair_stats()</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="de_pair_chisq.html">de_pair_chisq()</a></code> </p>
        </td>
        <td><p>Perform pairwise differential detection tests using Chi-Squared for a single pair of clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="de_pair_limma.html">de_pair_limma()</a></code> </p>
        </td>
        <td><p>Perform pairwise DE tests using limma for a single pair of clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="de_param.html">de_param()</a></code> </p>
        </td>
        <td><p>Set differential expression (DE) parameters for genes and clusters.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="de_score.html">de_score()</a></code> </p>
        </td>
        <td><p>reverse-compatibility wrapper around de_all_pairs_stats()</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="de_score_pairs.html">de_score_pairs()</a></code> </p>
        </td>
        <td><p>reverse-compatibility wrapper around de_selected_pairs_stats()</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="de_selected_pairs.html">de_selected_pairs()</a></code> </p>
        </td>
        <td><p>Perform pairwise differential gene expression tests between main pairs of clusters in parallel</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="de_stats_all_pairs.html">de_stats_all_pairs()</a></code> </p>
        </td>
        <td><p>Compute differential expression stats for all pairs of clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="de_stats_pair.html">de_stats_pair()</a></code> </p>
        </td>
        <td><p>Compute differential expression summary statistics based on a differential results data.frame and de_param().</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="de_stats_selected_pairs.html">de_stats_selected_pairs()</a></code> </p>
        </td>
        <td><p>Compute differential expression summary statistics for selected pairs of clusters based on de_param().</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="display_cl.html">display_cl()</a></code> </p>
        </td>
        <td><p>Display cluster plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="doubletFinder.html">doubletFinder()</a></code> </p>
        </td>
        <td><p>Doublet detection in single-cell RNA sequencing data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findVG.html">findVG()</a></code> </p>
        </td>
        <td><p>reverse-compatibility wrapper around find_vg()</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="find_low_quality_cl.html">find_low_quality_cl()</a></code> </p>
        </td>
        <td><p>Use DEGenes to identify clusters with few differentially expressed genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="find_vg.html">find_vg()</a></code> </p>
        </td>
        <td><p>Identify high variability genes using Brennecke's method</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_dispersion.html">gene_dispersion()</a></code> </p>
        </td>
        <td><p>Compute dispersion for each gene in a gene x sample matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_loess_fit.html">gene_loess_fit()</a></code> </p>
        </td>
        <td><p>Compute dispersion-mean Loess-fit for each gene in a gene x sample matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_loess_fit_z.html">gene_loess_fit_z()</a></code> </p>
        </td>
        <td><p>Compute dispersion z scores for the residuals of a loess fit</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_means.html">gene_means()</a></code> </p>
        </td>
        <td><p>Compute means for each gene in a gene x sample matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_vars.html">gene_vars()</a></code> </p>
        </td>
        <td><p>Compute variance for each gene in a gene x sample matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_z.html">gene_z()</a></code> </p>
        </td>
        <td><p>Compute dispersion z-scores for each gene in a gene x sample matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getTopMarkersByPropNew.html">getTopMarkersByPropNew()</a></code> </p>
        </td>
        <td><p>Find specific marker gene for each cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_cell.cl.co.ratio.html">get_cell.cl.co.ratio()</a></code> </p>
        </td>
        <td><p>Get cell co-clustering ratios</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_cl_df.html">get_cl_df()</a></code> </p>
        </td>
        <td><p>Generate an initial cl.df object based on cl</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_cl_mat.html">get_cl_mat()</a></code> </p>
        </td>
        <td><p>Generate a sparse matrix one-hot representation of clusters x samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_cl_means.html">get_cl_means()</a></code> </p>
        </td>
        <td><p>Compute cluster means for each row in a matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_cl_medians.html">get_cl_medians()</a></code> </p>
        </td>
        <td><p>Compute cluster medians for each row in a matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_cl_prop.html">get_cl_prop()</a></code> </p>
        </td>
        <td><p>Compute cluster proportions for each row in a matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_cl_sums.html">get_cl_sums()</a></code> </p>
        </td>
        <td><p>Compute cluster sums for each row in a matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_de_matrix.html">get_de_matrix()</a></code> </p>
        </td>
        <td><p>Generate a matrix of pairwise DE results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_eigen.html">get_eigen()</a></code> </p>
        </td>
        <td><p>Compute module eigengenes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_pairs.html">get_pairs()</a></code> </p>
        </td>
        <td><p>Convert underscore_separated pair names to a data.frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_pair_matrix.html">get_pair_matrix()</a></code> </p>
        </td>
        <td><p>Subset a matrix as a vector using row and column positions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_pair_matrix_coor.html">get_pair_matrix_coor()</a></code> </p>
        </td>
        <td><p>Convert matrix row/column positions to vector position</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="group_cl.html">group_cl()</a></code> </p>
        </td>
        <td><p>Assign clusters to a group</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="heatmap.3.html">heatmap.3()</a></code> </p>
        </td>
        <td><p>A modified version of heatmap.2 from the gplots package for cluster plotting</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="incrementHex.html">incrementHex()</a></code> </p>
        </td>
        <td><p>Adjust color vector</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="iter_clust.html">iter_clust()</a></code> </p>
        </td>
        <td><p>Iterative clustering algorithm for single cell RNAseq dataset</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="iter_consensus_clust.html">iter_consensus_clust()</a></code> </p>
        </td>
        <td><p>Iterative consensus clustering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="jaccard.html">jaccard()</a></code> </p>
        </td>
        <td><p>Compute jaccard distances between columns of a matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="jaccard_louvain.RANN.html">jaccard_louvain.RANN()</a></code> </p>
        </td>
        <td><p>Perform Jaccard/Louvain clustering based on RANN</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="jaccard_louvain.html">jaccard_louvain()</a></code> </p>
        </td>
        <td><p>Perform Jaccard/Louvain clustering using PhenoGraph clustering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="knn_jaccard.html">knn_jaccard()</a></code> </p>
        </td>
        <td><p>Compute a jaccard distance matrix based on a matrix of K-Nearest Neigbors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lm_matrix.html">lm_matrix()</a></code> </p>
        </td>
        <td><p>Perform a linear model fit on a matrix of data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lm_normalize.html">lm_normalize()</a></code> </p>
        </td>
        <td><p>Normalize using a linear model</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="logCPM.html">logCPM()</a></code> </p>
        </td>
        <td><p>Convert a matrix of raw counts to a matrix of log2(Counts per Million + 1) values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeColorsUnique.html">makeColorsUnique()</a></code> </p>
        </td>
        <td><p>Make a vector of unique colors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeLayerLabel.html">makeLayerLabel()</a></code> </p>
        </td>
        <td><p>Numeric layer for annotation object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeRegionLabel.html">makeRegionLabel()</a></code> </p>
        </td>
        <td><p>Split region from roi for annotation object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="map_by_cor.html">map_by_cor()</a></code> </p>
        </td>
        <td><p>Map samples to a training dataset by correlation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="map_cl_summary.html">map_cl_summary()</a></code> </p>
        </td>
        <td><p>Map a dataset to a reference, and compare existing cluster calls to the reference comparison</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="map_cv.html">map_cv()</a></code> </p>
        </td>
        <td><p>Run a single round of cross-validation of cluster mapping using a subset of marker genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="map_sampling.html">map_sampling()</a></code> </p>
        </td>
        <td><p>Perform bootstrapped mapping using a fraction of provided marker genes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="match_cl.html">match_cl()</a></code> </p>
        </td>
        <td><p>Compute correlations of clusters to a reference set, and get the best-correlated reference clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="merge_cl.html">merge_cl()</a></code> </p>
        </td>
        <td><p>Merge clusters based on pairwise differential expressed genes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="onestep_clust.html">onestep_clust()</a></code> </p>
        </td>
        <td><p>One round of clustering in the iteractive clustering pipeline</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pair_cor.html">pair_cor()</a></code> </p>
        </td>
        <td><p>Compute correlations between each matching row or column of two matrices</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pass_louvain.html">pass_louvain()</a></code> </p>
        </td>
        <td><p>Test louvain modularity scores based on an adjacency matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="phenograph.html">phenograph()</a></code> </p>
        </td>
        <td><p>Perform PhenoGraph clustering on a matrix of data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="phenograph_jaccard_coeff.html">phenograph_jaccard_coeff()</a></code> </p>
        </td>
        <td><p>Compute Jaccard Coefficients from a K nearest neighbors matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_low_qc.html">plot_low_qc()</a></code> </p>
        </td>
        <td><p>Plot low-quality clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_vg.html">plot_vg()</a></code> </p>
        </td>
        <td><p>Generate gene variance plots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="predict_annotate_cor.html">predict_annotate_cor()</a></code> </p>
        </td>
        <td><p>Predict annotations by cluster correlation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rd_PCA.html">rd_PCA()</a></code> </p>
        </td>
        <td><p>Reduce dimensionality using PCA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rd_WGCNA.html">rd_WGCNA()</a></code> </p>
        </td>
        <td><p>Compute reduced dimensions using WGCNA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refine_cl.html">refine_cl()</a></code> </p>
        </td>
        <td><p>Refine clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="renameAndOrderClusters.html">renameAndOrderClusters()</a></code> </p>
        </td>
        <td><p>Rename clusters using genes and metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reorder_cl.html">reorder_cl()</a></code> </p>
        </td>
        <td><p>Reorder cluster based on hiearchical clustering of mean values from the input data matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reorder_factor.html">reorder_factor()</a></code> </p>
        </td>
        <td><p>Reorder factors of one annotation to match another</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rescale_samples.html">rescale_samples()</a></code> </p>
        </td>
        <td><p>Rescale a gene x sample matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_consensus_clust.html">run_consensus_clust()</a></code> </p>
        </td>
        <td><p>Wrapper function to repeatively run clustering on subsampled cells and infer consensus clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sample_cells.html">sample_cells()</a></code> </p>
        </td>
        <td><p>Downsample cells from each cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="score_gene_mod.html">score_gene_mod()</a></code> </p>
        </td>
        <td><p>Compute scores for a set of gene modules</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="select_markers.html">select_markers()</a></code> </p>
        </td>
        <td><p>Select a set of marker genes based on differential gene expression results.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="set_pair_matrix.html">set_pair_matrix()</a></code> </p>
        </td>
        <td><p>Update a matrix with values from a 1d vector using row and column positions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sparse_cor.html">sparse_cor()</a></code> </p>
        </td>
        <td><p>Compute correlation scores for columns of a sparse matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="updateSampDat.html">updateSampDat()</a></code> </p>
        </td>
        <td><p>Update cluster names in annotation data frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vec_chisq_test.html">vec_chisq_test()</a></code> </p>
        </td>
        <td><p>Vectorized Chi-squared tests for differential gene detection</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zizhen Yao, Lucas Graybuck, Trygve Bakken.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

